fl4health/parameter_exchange/sparse_coo_parameter_exchanger.py

```diff
@@ -0,0 +1,154 @@
+            score_gen_function (Callable[..., Dict[str, Tensor]]): Function that is responsible for
+            generating a score for every parameter inside a model in order to facilitate parameter selection.
+
+            In most cases, this function takes as inputs a current model and an initial model,
+            and it returns a dictionary that maps the name of each of the current model's tensors to
+            another tensor which contains the parameter scores.
+        """
+        assert 0 < sparsity_level <= 1
+        self.set_sparsity_level(sparsity_level)
+        self.parameter_packer: SparseCooParameterPacker = SparseCooParameterPacker()
+        self.set_score_gen_function(score_gen_function)
+
+    def set_sparsity_level(self, sparsity_level: float) -> None:
+        """This method exists to allow for setting new sparsity levels throughout training."""
+        self.sparsity_level = sparsity_level
+
+    def set_score_gen_function(self, score_gen_function: Callable[..., Dict[str, Tensor]]) -> None:
+        self.score_gen_function = score_gen_function
+
+    def generate_parameter_scores(self, model: nn.Module, initial_model: Optional[nn.Module]) -> Dict[str, Tensor]:
+        """Calling the score generating function to produce parameter scores."""
+        return self.score_gen_function(model, initial_model)
+
+    def create_masks(self, model: nn.Module, initial_model: Optional[nn.Module]) -> Dict[str, Tensor]:
+        """
+        Produce masks for model's tensors.
+
+        More precisely, this method first leverage a score generating function
+        to generate scores for all parameters of model.
+
+        Next, these scores are used to produce masks.
+        A threshold is determined according to the desired sparsity level,
+        then parameters whose scores are less than this threshold are assigned a mask value of 0,
+        while parameters whose scores are greater than or equal to this threshold are
+        assigned a mask value of 1.
+
+        Args:
+            model (nn.Module): Current model.
+            initial_model (nn.Module): Initial model.
+
+        Returns:
+            Dict[str, Tensor]: Dictionary that maps the name of each tensor to its corresponding mask.
+        """
+        all_parameter_scores = self.generate_parameter_scores(model, initial_model)
+        all_scores = torch.cat([val.flatten() for _, val in all_parameter_scores.items()])
+        # Sorting all scores and determining the threshold.
+        sorted_scores, _ = torch.sort(all_scores, descending=True)
+        top_scores = sorted_scores[: math.ceil(len(sorted_scores) * self.sparsity_level)]
+        score_threshold = top_scores[-1].item()
+
+        names_to_masks = {}
+        for name, param_scores in all_parameter_scores.items():
+            # Use score_threshold to produce masks.
+            threshold_result = torch.where(param_scores >= score_threshold, input=param_scores, other=0)
```

Comment:
I might be wrong, but I think `threshold_result` is the same tensor as 
```
sparse_tensor = torch.multiply(model_tensor, tensor_mask)
```
that you're computing below, which is what you really want anyway. Rather than forming the names_to_masks dictionary here and then multiplying by the mask to get the sparse_tensor, maybe we can just form a names_to_sparse_tensor_dictionary here? You can still not add them to the dictionary using `if not (threshold_result == 0).all():` 

Correct me if I'm wrong though.

Reply:
`threshold_result` here is not exactly the result obtained after multiplying `model_tensor` by `tensor_mask`,  since  `torch.where()` is applied on the parameters scores tensor rather than `model_tensor` itself. But you are right that we can rewrite this part to avoid having to create a names_to_masks dictionary, so I'll reimplement it this way.